src/bridge/cockpitdbuscache.c

```diff
@@ -150,25 +152,52 @@ barrier_progress (CockpitDBusCache *self)
   BarrierData *barrier;
   BatchData *batch;
 
-  batch = g_queue_peek_head (self->batches);
-
-  for (;;)
+  if (self->barrier_progressing)
     {
-      barrier = g_queue_peek_head (self->barriers);
-      if (!barrier)
-        return;
-
-      /*
-       * If there is a batch being processed, we must block
-       * barriers that have an equal or later batch number.
+      /* It might happen that barrier_progress is called
+       * recursively, while one of the barrier callbacks is
+       * running.  We need to detect this and avoid calling the
+       * next barrier callback while there is already one
+       * running.  Allowing it might violate the ordering
+       * guarantees if the running callback is doing more work
+       * after causing the recursive call to barrier_progress.
+       *
+       * Concretely, process_interfaces below can cause
+       * recursive calls in the middle of its work when it
+       * processes multiple interfaces.
        */
-      if (batch && batch->number <= barrier->number)
-        return;
 
-      g_queue_pop_head (self->barriers);
-      (barrier->callback) (self, barrier->user_data);
-      g_slice_free (BarrierData, barrier);
+      self->barrier_more_progress = TRUE;
```

Comment:
I suggest using a plain counter and decrementing that in the `barrier_progress` loop. This corresponds to a semaphore-like pattern instead of a mutex.